### ISSUE TITLE
Delegate creating the GradSampler to PruningModifiers

### ISSUE THREAD

**Issue by eldarkurtic** [neuralmagic/sparseml#823]
Currently, the GradSampler is instantiated in https://github.com/neuralmagic/sparseml/blob/bd64454b3ae7b40bccd9057957d5910f45eda4cb/src/sparseml/transformers/sparsification/trainer.py#L158

We should consider delegating this process to the PruningModifier class that makes use of it, because:
- now, all pruning modifiers will instantiate it, even though they might not need it (for example MagnitudePruningModifier)
- currently `RecipeManagerTrainerInterface` class is creating it with `self._mfac_data_loader()` and `self._mfac_loss_function`, but `MFAC` is not the only pruner that might use the GradSampler, so we probably shouldn't call it `mfac`
- by delegating this process to the PruningModifier class we also have the opportunity to modify hyperparams when dataloader and loss function are created
- this is very important for `MFAC` and `oBERT` pruning modifiers because the batch-size for GradSampler is an important hyperparam to be configurable from the recipe
- currently this is hardcoded to use `train_batch_size/2` https://github.com/neuralmagic/sparseml/blob/bd64454b3ae7b40bccd9057957d5910f45eda4cb/src/sparseml/transformers/sparsification/trainer.py#L667 which will be a problem when running experiments where `train_batch_size` is huge (for example 256); it's going to be very hard to fit that on a GPU
